packages/build/src/plugins/compatibility.test.ts

```diff
@@ -1,226 +1,319 @@
+        conditions: [{ type: 'siteDependencies', condition: { next: '<10.0.9' } }],
+      },
+    ]
+
+    const { version: version1 } = await getExpectedVersion({
+      versions,
+      nodeVersion: '20.0.0',
+      packageJson: { dependencies: { next: '14.0.0' } },
+      buildDir: '/some/path',
+      pinnedVersion: '4',
+      systemLog: noopSystemLog,
+    })
+    expect(version1).toBe('5.0.0-beta.1')
+
+    // out of range
+    const { version: version2 } = await getExpectedVersion({
+      versions,
+      nodeVersion: '20.0.0',
+      packageJson: { dependencies: { next: '13.0.0' } },
+      buildDir: '/some/path',
+      pinnedVersion: '4',
+      systemLog: noopSystemLog,
+    })
+    expect(version2).toBe('4.41.2')
   })
-  expect(version).toBe('5.0.0-beta.1')
-})
 
-test('`getExpectedVersion` should retrieve the plugin based on conditions and feature flag due to pinned version', async () => {
-  const versions: PluginVersion[] = [
-    {
-      version: '5.0.0-beta.1',
-      conditions: [
-        { type: 'nodeVersion', condition: '>= 18.0.0' },
-        { type: 'siteDependencies', condition: { next: '>=13.5.0' } },
-      ],
-      overridePinnedVersion: '>=4.0.0',
-    },
-    { version: '4.41.2', conditions: [] },
-    {
-      version: '3.9.2',
-      conditions: [{ type: 'siteDependencies', condition: { next: '<10.0.9' } }],
-    },
-  ]
-
-  const { version: version1 } = await getExpectedVersion({
-    versions,
-    nodeVersion: '20.0.0',
-    packageJson: { dependencies: { next: '14.0.0' } },
-    buildDir: '/some/path',
-    pinnedVersion: '4',
+  test('matches the first entry that satisfies the constraints, even if it also matches another entry further down with more specific constraints', async () => {
+    const versions: PluginVersion[] = [
+      { version: '4.41.2', conditions: [] },
+      {
+        version: '5.0.0-beta.1',
+        conditions: [
+          { type: 'nodeVersion', condition: '>= 18.0.0' },
+          { type: 'siteDependencies', condition: { next: '>=13.5.0' } },
+        ],
+        overridePinnedVersion: '>=4.0.0',
+      },
+      {
+        version: '3.9.2',
+        conditions: [{ type: 'siteDependencies', condition: { next: '<10.0.9' } }],
+      },
+    ]
+
+    const { version } = await getExpectedVersion({
+      versions,
+      nodeVersion: '20.0.0',
+      packageJson: { dependencies: { next: '14.0.0' } },
+      buildDir: '/some/path',
+      pinnedVersion: '4',
+      systemLog: noopSystemLog,
+    })
+    expect(version).toBe('4.41.2')
   })
-  expect(version1).toBe('5.0.0-beta.1')
-
-  // out of range
-  const { version: version2 } = await getExpectedVersion({
-    versions,
-    nodeVersion: '20.0.0',
-    packageJson: { dependencies: { next: '13.0.0' } },
-    buildDir: '/some/path',
-    pinnedVersion: '4',
+
+  test('if no pinned version is set, it matches the first version regardless of whether its requirements match the conditions (legacy behavior)', async () => {
```

Comment:
The diff makes it a bit hard to see (sorry), but this is one of the new tests.

Reply:
For other reviewers - https://github.com/netlify/build/pull/5575/files?w=1 (hide whitespace changes in diff) makes this diff much easier to parse

Reply:
> For other reviewers - https://github.com/netlify/build/pull/5575/files?w=1 (hide whitespace changes in diff)

TIL, ty!